Sprint 17. Переход c Creator на Koin

amdoit-git opened the pull request.

amdoit-git committed:
Поиск сделан из потока который находится в фоне и не умигает никогда

amdoit-git committed:
ITunesRepositoryImpl реализует не только ITunesRepository, но и Thread. Так удалось избавиться от вызова конструктора потока внутри конструктора репозитория реализующего интерфейс

amdoit-git committed:
Удален Creator.kt
Экземпяры классов передаются в конструкторы через параметры, а не создаются внутри
Зависимости внедряются через Koin

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/common/data/SearchHistory.kt:
```diff
@@ -11,13 +11,17 @@ object SearchHistory {
+    private lateinit var sharedPrefs: SharedPreferences
+    private lateinit var gson: Gson
 
     fun setSharedPreferences(sharedPrefs: SharedPreferences) {
         this.sharedPrefs = sharedPrefs
     }
 
+    fun setGson(gson: Gson) {
+        this.gson = gson
+    }
```
Зависимости нужно передавать через конструктор, объект нужно переделать в класс и добавить его в граф

amdoit-git replied:
Удалил SearchHistory перенеся его код в репозиторий. Переделал MediaPlayerService в класс, который получает плеер через конструктор, а не публичный метод

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/di/UIModule.kt:
```diff
@@ -0,0 +1,37 @@
+        TrackAdapter(onTrackClick, onButtonClick, scrollListToTop, get())
+    }
+
+    factory<MutableList<Track>> {
+        mutableListOf()
+    }
```
Выглядит лишним, просто создание пустого мьютабельного списка, это в графе не нужно

amdoit-git replied:
Перенес mutableListOf() в конструктор адаптера

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/di/UIModule.kt:
```diff
@@ -0,0 +1,37 @@
+
+val uiModule = module {
+
+    factory<ActivityPlayerScreenBinding> { (layoutInflater: LayoutInflater) ->
+        ActivityPlayerScreenBinding.inflate(layoutInflater)
+    }
+
+    factory<ActivitySettingsBinding> { (layoutInflater: LayoutInflater) ->
+        ActivitySettingsBinding.inflate(layoutInflater)
+    }
+
+    factory<ActivitySearchBinding> { (layoutInflater: LayoutInflater) ->
+        ActivitySearchBinding.inflate(layoutInflater)
+    }
+
+    factory<ActivityMainBinding> { (layoutInflater: LayoutInflater) ->
+        ActivityMainBinding.inflate(layoutInflater)
+    }
```
В целом биндинги тоже лишние, для их создания ничего кроме `LayoutInflater` не понадобится никогда, да и один биндинг как правило на 1 активити/фрагмент, поэтому выносить это в граф не даёт никаких преимуществ

amdoit-git replied:
Перенес создание биндингов в Активити

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/main/presentation/MainViewModel.kt:
```diff
@@ -1,27 +1,28 @@
-            return MainViewModel(application) as T
-        }
-    }
+//    class Factory(private val application: Application) :
```
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

amdoit-git replied:
Удалил закомментированный код

amdoit-git committed:
- Удалил закомментированный код
- Перенес создание биндингов в Активити
- Перенес mutableListOf() в конструктор адаптера
- Удалил SearchHistory перенеся его код в репозиторий. Переделал MediaPlayerService в класс, который получает плеер через конструктор, а не публичный метод

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/di/DataModule.kt:
```diff
@@ -22,6 +23,10 @@ val dataModule = module {
         MediaPlayer()
     }
 
+    single<MediaPlayerService> {
+        MediaPlayerService(mediaPlayer = get())
+    }
```
Здесь нельзя использовать `single`, т.к. инстанс `MediaPlayer` нужно пересоздавать на каждый новый трек, иначе будет креш

amdoit-git replied:
Это почему будет креш?) Он позволяет устанавливать новый источник данных без пересоздания

amdoit-git replied:
Раньше он был просто в Object  и создавался при создании Object. Все работает. Проверено

ArturNurtdinov replied:
Поправлюсь - креш будет, если ты вызываешь `release` у `MediaPlayer` - тогда при следующем `prepare` выбросится ошибка. Но если `release` не вызывать, то вполне можно менять источник данных, но тут проблема с тем, что плеер будет всё время храниться в памяти и потреблять ресурсы

amdoit-git replied:
Да. Зато не будут тратится ресурсы на его постоянное пересоздание. А учитывая что проигрывание музыки - основной функционал плеер будет работать большую часть времени. Вообще конечно нужно тест какой то сделать и выяснить как лучше с точки зрения производительности. В общем single я специально сделал. И на этот счёт в заданиях никаких указаний не было
